File: solutions/cpp/Median_sorted_circular_linked_list.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>
#include <random>
#include <stdexcept>

#include "./Linked_list_prototype.h"

using std::cout;
using std::default_random_engine;
using std::endl;
using std::exception;
using std::length_error;
using std::make_shared;
using std::random_device;
using std::uniform_int_distribution;

// @include
double FindMedianSortedCircularLinkedList(
    const shared_ptr<ListNode<int>>& arbitrary_node) {
  // Checks if all nodes are identical and identifies the first smallest node.
  auto iter = arbitrary_node, first_smallest_node = arbitrary_node;
  int n = 0;
  do {
    if (iter->data > iter->next->data) {
      first_smallest_node = iter->next;
    }
    ++n, iter = iter->next;
  } while (iter != arbitrary_node);

  // Advances to the middle of the list.
  for (int i = 0; i < ((n - 1) / 2); ++i) {
    first_smallest_node = first_smallest_node->next;
  }
  return n % 2 ? first_smallest_node->data
               : 0.5 * (first_smallest_node->data +
                        first_smallest_node->next->data);
}
// @exclude

void SmallTest() {
  shared_ptr<ListNode<int>> L = make_shared<ListNode<int>>(
      ListNode<int>{0, make_shared<ListNode<int>>(ListNode<int>{
                           2, make_shared<ListNode<int>>(ListNode<int>{
                                  2, make_shared<ListNode<int>>(
                                         ListNode<int>{2, nullptr})})})});
  L->next->next->next->next = L;
  cout << FindMedianSortedCircularLinkedList(L->next->next) << endl;
  assert(2 == FindMedianSortedCircularLinkedList(L->next->next));
}

int main(int argc, char* argv[]) {
  SmallTest();
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    int n;
    if (argc == 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(1, 1000);
      n = dis(gen);
    }
    shared_ptr<ListNode<int>> head;
    for (int i = n; i >= 0; --i) {
      auto curr = make_shared<ListNode<int>>(ListNode<int>{i, nullptr});
      curr->next = head;
      head = curr;
    }
    shared_ptr<ListNode<int>> curr = head;
    if (curr != shared_ptr<ListNode<int>>(nullptr)) {
      while (curr->next != shared_ptr<ListNode<int>>(nullptr)) {
        curr = curr->next;
      }
      curr->next = head;  // make the list as a circular list.
    }
    double res = FindMedianSortedCircularLinkedList(head->next);
    cout << res << endl;
    assert(res == 0.5 * n);
  }

  // Test identical list.
  shared_ptr<ListNode<int>> head;
  for (int i = 0; i < 10; ++i) {
    auto curr = make_shared<ListNode<int>>(ListNode<int>{5, nullptr});
    curr->next = head;
    head = curr;
  }
  shared_ptr<ListNode<int>> curr = head;
  if (curr != shared_ptr<ListNode<int>>(nullptr)) {
    while (curr->next != nullptr) {
      curr = curr->next;
    }
    curr->next = head;  // make the list as a circular list.
  }
  assert(5 == FindMedianSortedCircularLinkedList(head));
  return 0;
}

File: solutions/python/median_sorted_circular_linked_list.py
import sys
import random
from linked_list_prototype import ListNode


# @include
def find_median_sorted_circular_linked_list(arbitrary_node):
    # Checks if all nodes are identical and identifies the first smallest node.
    it = arbitrary_node
    first_smallest_node = arbitrary_node
    n = 0
    while True:
        n += 1
        it = it.next
        if first_smallest_node.data <= it.data:
            # Now first_smallest_node points to the largest element.
            first_smallest_node = it
        if it is arbitrary_node:
            break
    # first_smallest_node's next is the first smallest node.
    first_smallest_node = first_smallest_node.next

    # Advances to the middle of the list.
    for i in range((n - 1) // 2):
        first_smallest_node = first_smallest_node.next

    if n % 2:
        return first_smallest_node.data
    else:
        return 0.5 * (first_smallest_node.data + first_smallest_node.next.data)
# @exclude


def main():
    for _ in range(1000):
        n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 1000)
        head = None
        for i in reversed(range(n + 1)):
            curr = ListNode(i, head)
            head = curr
        curr = head
        if curr:
            while curr.next:
                curr = curr.next
            curr.next = head  # make the list as a circular list.
        res = find_median_sorted_circular_linked_list(head.next)
        print(res)
        assert res == 0.5 * n

    # Test identical list.
    head = None
    for i in range(10):
        curr = ListNode(5, head)
        head = curr
    curr = head
    if curr:
        while curr.next:
            curr = curr.next
        curr.next = head  # make the list as a circular list.
    assert 5 == find_median_sorted_circular_linked_list(head)


if __name__ == '__main__':
    main()
